internal/driver/tagroot.go

```diff
@@ -0,0 +1,120 @@
+package driver
+
+import (
+	"strings"
+
+	"github.com/google/pprof/internal/measurement"
+	"github.com/google/pprof/profile"
+)
+
+// addLabelNodes adds pseudo stack frames "label:value" to each Sample with
+// labels matching the supplied keys.
+//
+// rootKeys adds frames at the root of the callgraph (first key becomes new root).
+// leafKeys adds frames at the leaf of the callgraph (last key becomes new leaf).
+//
+// Returns whether there were matches found for the label keys.
+func addLabelNodes(p *profile.Profile, rootKeys, leafKeys []string, outputUnit string) (rootm, leafm bool) {
+	// Find where to insert the new locations and functions at the end of
+	// their ID spaces.
+	var maxLocID uint64
+	var maxFunctionID uint64
+	for _, loc := range p.Location {
+		if loc.ID > maxLocID {
+			maxLocID = loc.ID
+		}
+	}
+	for _, f := range p.Function {
+		if f.ID > maxFunctionID {
+			maxFunctionID = f.ID
+		}
+	}
+	nextLocID := maxLocID + 1
+	nextFuncID := maxFunctionID + 1
+
+	// Intern the new locations and functions we are generating.
+	locsByName := map[string]*profile.Location{}
+	functionsByName := map[string]*profile.Function{}
+
+	internFunction := func(functionName string) *profile.Function {
+		function, found := functionsByName[functionName]
+		if found {
+			return function
+		}
+		function = &profile.Function{
+			ID:   nextFuncID,
+			Name: functionName,
+		}
+		nextFuncID++
+		p.Function = append(p.Function, function)
+		functionsByName[functionName] = function
+		return function
+	}
+
+	internLocation := func(functionName string, function *profile.Function) *profile.Location {
+		location, found := locsByName[functionName]
+		if found {
+			return location
+		}
+		location = &profile.Location{
+			ID: nextLocID,
+			Line: []profile.Line{
+				profile.Line{
+					Function: function,
+				},
+			},
+		}
+		nextLocID++
+		p.Location = append(p.Location, location)
+		locsByName[functionName] = location
+		return location
+	}
+
+	for _, s := range p.Sample {
+		rootsToAdd := []*profile.Location{}
+		leavesToAdd := []*profile.Location{}
+		for i := range rootKeys {
```

Comment:
> the body of the rootKeys and leafKeys functions is the same

They're very similar... but they're "different enough" in which variables they mutate that I'm not sure it'd be clearer to pull out a function?

I'd have to pass a lot of pointers to variables like `leafm` vs `rootm`, `rootsToAdd` vs `leavesToAdd`, or have a lot of multiple return values (which I'm not sure would be clearer?) That said, I'm mostly a Java programmer day-to-day and maybe it's totally fine and normal to pass pointers in Go so that e.g. slices can be mutated. Is that normal, like to pass a `*bool`?

Maybe I should give it a go; after all, I did have some copy/paste errors here. I'll think a bit and try some more variations.

> sort the keys so that the order is deterministic

Yeah this is a great idea, I had forgotten I'm dealing with hashtables here.

Reply:
> sort the keys so that the order is deterministic

Hmm, I think we aren't iterating a hashtable here? (but I might be missing something):
- We loop over `rootKeys, leafKeys` which are `[]string` passed from command line flags
- We look up the map using the keys from `rootKeys, leafKeys` in `formatLabelValues`.

Reply:
> the body of the rootKeys and leafKeys functions is the same

I extracted a function, PTAL. I wonder if I might have too many closures here, there's a few intern tables and outputUnit being passed through closure. If this is getting a bit many closures (I'm not sure of Go style) then maybe it's time to make a class.

Reply:
I don't seem to see the changes - not pushed yet?